ldm/invoke/generator/diffusers_pipeline.py

```diff
@@ -404,6 +404,10 @@ def generate_latents_from_embeddings(self, latents: torch.Tensor, timesteps,
                                         total_step_count=len(timesteps),
                                         additional_guidance=additional_guidance)
                 latents = step_output.prev_sample
+
+                if conditioning_data.threshold:
+                    latents = self.invokeai_diffuser.do_thresholding(conditioning_data.threshold.threshold, conditioning_data.threshold.warmup, latents, batched_t, i, len(timesteps))
```

Comment:
How does this benefit from being here instead of in the InvokeAIDiffuserComponent?

The pipeline class is more than a handful on its own, and I'd like to avoid adding more special-case branches for every effect we think up here.

Reply:
When I moved this a layer lower, it simply didn't work. Maybe this comes down to me not knowing much about pytorch, but it seemed like it had to be here to get things to work. (Same with the symmetry I was toying with.)

Reply:
@damian0815 Any ideas as to why this is the case and why applying thresholding at a level down doesn't work?

Reply:
@keturn Another option is having a generic `apply_transformations_to_latents` at the `InvokeAIDiffuserComponent` level and we simply call that with all parameters here and it determines what to do - thresholding, symmetry, color shifting, etc. Thoughts?

Reply:
> ^^ Not to scope creep, but my thought here is always that it's nice having named callbacks, i.e. being able to register hooks / filters that can run later in a particular spot in the code, and this is a great example of where a system like that would be useful.

I also wonder how this would interact with nodes, if at all. I don't know whether or not nodes will let us operate on per-step latents.

Reply:
> @keturn Another option is having a generic `apply_transformations_to_latents` at the `InvokeAIDiffuserComponent` level and we simply call that with all parameters here and it determines what to do - thresholding, symmetry, color shifting, etc. Thoughts?

this would be my preferred design too

Reply:
> @damian0815 Any ideas as to why this is the case and why applying thresholding at a level down doesn't work?

@JPPhoto without knowing what `step()` does i can't say. i suspect however (and i guess @keturn can confirm) that `step()` does some stuff `a`, calls through to `InvokeAIDiffusionComponent.do_diffusion_step()`, and then does some more stuff `b`, and the thresholding code for some reason needs to happen either completely before `a` or entirely after `b`. i could be wrong though.

Reply:
@keturn @damian0815 I believe I've addressed your concerns with the latest changes. This should also let us push new features to `InvokeAIDiffuserComponent` like symmetry without affecting `StableDiffusionGeneratorPipeline`. Please have a look.

Reply:
Looks good, my only thought is that it might be worth going the whole hog of refactorisationg and further busting out the `latents = self._threshold(postprocessing_settings.threshold, postprocessing_settings.warmup, latents, percent_through)` line to its own function, so that the call tree would be like `generate_latents_from_embeddings` → `do_latent_postprocessing()` → `apply_threshold()`

Reply:
Wouldn't `apply_threshold` just be renaming `_threshold` then? Or are you thinking of a different signature?